File: 98-validate-binary-search-tree/98-validate-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        
       
        inf = float('inf') 
        st = [ [root, -inf , inf] ] 
        while st: 

            node , lw , up = st.pop()
            if not node: continue 
            v = node.val
            if v <= lw or v>= up: 
                return False 
            st.append([node.right , v , up])
            st.append([node.left , lw , v ])

        return True 

File: 98-validate-binary-search-tree/98-validate-binary-search-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean isValidBST(TreeNode root) {
        return isValidBST(root, Long.MIN_VALUE , Long.MAX_VALUE);
    }
    public boolean isValidBST(TreeNode root, long mnv ,long mxv){
        if(root==null) return true ; 
        if(root.val >= mxv || root.val <= mnv ) return false ; 
        return isValidBST(root.left , mnv , root.val) && isValidBST( root.right , root.val , mxv) ; 
    }
        
        
    
}